### ISSUE TITLE
How to whole-archive a library in cgo?

### ISSUE THREAD

**Issue by tanjianfeng** [bazelbuild/rules_go#2201]
### What version of rules_go are you using?

0.19.3

### What version of gazelle are you using?

0.18.1

### What version of Bazel are you using?

0.28.1

### Does this issue reproduce with the latest releases of all the above?

yes.

### What operating system and processor architecture are you using?

Linux/amd64

### What did you do?

We have a c library (https://github.com/tanjianfeng/clib).

clib.c
```
#include <stdio.h>

static void __attribute__((constructor(101)))
clib_init(void)
{
	printf("%s, hello\n", __func__);
}

extern int
clib_hello(void)
{
	printf("%s, hello\n", __func__);
}
```

helper.c
```
#include <stdio.h>

static void __attribute__((constructor(101)))
helper_init(void)
{
	printf("%s, hello\n", __func__);
}
```

In a go project (https://github.com/tanjianfeng/bazel-examples/tree/master/cgo), we only use the clib_hello() in clib.c. To compile it:
`
bazel build :main -s
`

After compiling, we cannot see the symbol helper_init(), seems such symbols (not used) get dropped.

I doubt it's related to -ffunction-sections and -fdata-sections in the compile command:

```
cd /root/.cache/bazel/_bazel_root/f6242c61253b199c8f5d9e7a006d416c/execroot/__main__ && \
  exec env - \
    CC=/usr/bin/gcc \
    CGO_ENABLED=1 \
    GOARCH=amd64 \
    GOOS=linux \
    GOROOT=external/go_sdk \
    GOROOT_FINAL=GOROOT \
    PATH=/usr/bin:/bin \
  bazel-out/host/bin/external/go_sdk/builder compilepkg -sdk external/go_sdk -installsuffix linux_amd64 -src main.go -p main -package_list bazel-out/host/bin/external/go_sdk/packages.txt -o bazel-out/k8-fastbuild/bin/linux_amd64_stripped/main%/main.a -gcflags '' -asmflags '' -cppflags '-iquote external/libclib -iquote bazel-out/k8-fastbuild/bin/external/libclib' -cflags '-U_FORTIFY_SOURCE -fstack-protector -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -O2 -D_FORTIFY_SOURCE=1 -DNDEBUG -ffunction-sections -fdata-sections -fno-canonical-system-headers -Wno-builtin-macro-redefined -D__DATE__="redacted" -D__TIMESTAMP__="redacted" -D__TIME__="redacted" -fPIC' -cxxflags '-U_FORTIFY_SOURCE -fstack-protector -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -O2 -D_FORTIFY_SOURCE=1 -DNDEBUG -ffunction-sections -fdata-sections -std=c++0x -fno-canonical-system-headers -Wno-builtin-macro-redefined -D__DATE__="redacted" -D__TIMESTAMP__="redacted" -D__TIME__="redacted" -fPIC' -objcflags '-U_FORTIFY_SOURCE -fstack-protector -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -O2 -D_FORTIFY_SOURCE=1 -DNDEBUG -ffunction-sections -fdata-sections -fno-canonical-system-headers -Wno-builtin-macro-redefined -D__DATE__="redacted" -D__TIMESTAMP__="redacted" -D__TIME__="redacted" -fPIC' -objcxxflags '-U_FORTIFY_SOURCE -fstack-protector -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -O2 -D_FORTIFY_SOURCE=1 -DNDEBUG -ffunction-sections -fdata-sections -std=c++0x -fno-canonical-system-headers -Wno-builtin-macro-redefined -D__DATE__="redacted" -D__TIMESTAMP__="redacted" -D__TIME__="redacted" -fPIC' -ldflags 'bazel-out/k8-fastbuild/bin/external/libclib/libclib.a -fuse-ld=gold -Wl,-no-as-needed -Wl,-z,relro,-z,now -B/usr/bin -pass-exit-codes -lstdc++ -lm')
SUBCOMMAND: # //:main [action 'GoLink linux_amd64_stripped/main'

```

### What did you expect to see?

I'd like to see all symbols get compiled into the final binary:
nm bazel-bin/linux_amd64_debug/main | grep helper_init

### What did you see instead?
We cannot this helper_init() in the final binary.

**Comment by tanjianfeng** [comment]
I can use below change to work around this issue, but does this seem an issue to you? And does this worth a pull request?

```
diff --git a/go/private/rules/cgo.bzl b/go/private/rules/cgo.bzl
index 6fe41447..ab8adaab 100644
--- a/go/private/rules/cgo.bzl
+++ b/go/private/rules/cgo.bzl
@@ -161,7 +161,9 @@ def cgo_configure(go, srcs, cdeps, cppopts, copts, cxxopts, clinkopts):
                     clinkopts.extend(["-L", lib.dirname, "-l", libname])
                     inputs_direct.append(lib)
                 else:
+                    lib_opts.append("-Wl,--whole-archive")
                     lib_opts.append(lib.path)
+                    lib_opts.append("-Wl,--no-whole-archive")
             clinkopts.extend(cc_link_flags(d))

         elif hasattr(d, "objc"):
```

**Comment by jayconrod** [comment]
I think this is a duplicate of #1486.

As described in that issue passing `-Wl,--whole-archive` and `-Wl,--no-whole-archive` fixes this, but those should only be passed for `cc_libraries` that specifically request that. Bazel exposes that in the `CcInfo` provider now, so the work is unblocked.
